resolver-dns/src/main/java/io/netty/resolver/dns/DnsResolveContext.java

```diff
@@ -465,13 +521,48 @@ private boolean handleRedirect(
                         continue;
                     }
 
-                    nameServers.add(new InetSocketAddress(resolved, parent.dnsRedirectPort(resolved)));
-                    addNameServerToCache(authoritativeNameServer, resolved, r.timeToLive());
+                    final InetSocketAddress socketAddress = parent.newRedirectServerAddress(resolved);
+                    resolvedNameServers.add(socketAddress);
+
+                    System.err.println(socketAddress);
+                    addNameServerToCache(authoritativeNameServer, socketAddress, r.timeToLive());
                 }
+                final Set<InetSocketAddress> unresolvedNameServers;
+                if (serverNames.numUnhandled() > 0) {
+                    unresolvedNameServers = new LinkedHashSet<InetSocketAddress>(serverNames.numUnhandled());
+
+                    // Process all unresolved nameservers as well.
+                    for (;;) {
+                        final AuthoritativeNameServer authoritativeNameServer = serverNames.handleNext();
+                        if (authoritativeNameServer == null) {
+                            break;
+                        }
+                        // These will be resolved on the fly if needed.
+                        final InetSocketAddress unresolved = InetSocketAddress.createUnresolved(
+                                authoritativeNameServer.nsName, DefaultDnsServerAddressStreamProvider.DNS_PORT);
+                        unresolvedNameServers.add(unresolved);
 
-                if (!nameServers.isEmpty()) {
-                    query(parent.uncachedRedirectDnsServerStream(nameServers), 0, question,
-                          queryLifecycleObserver.queryRedirected(unmodifiableList(nameServers)), promise, null);
+                        addNameServerToCache(authoritativeNameServer, unresolved, authoritativeNameServer.timeToLive());
+                    }
+                } else {
+                    unresolvedNameServers = Collections.emptySet();
+                }
+
+                if (!resolvedNameServers.isEmpty() || !unresolvedNameServers.isEmpty()) {
+                    System.err.println(resolvedNameServers);
+
+                    List<InetSocketAddress> addresses = new ArrayList<InetSocketAddress>(
+                            resolvedNameServers.size() + unresolvedNameServers.size());
+                    addresses.addAll(resolvedNameServers);
+                    addresses.addAll(unresolvedNameServers);
```

Comment:
This logic that concats `resolvedNameServers` and `unresolvedNameServers` could be moved to the default implementation of `uncachedRedirectDnsServerList()`? By doing so, a custom `uncachedRedirectDnsServerList()` implementation may be able to construct the server list more efficiently.

Reply:
Good idea